### ISSUE TITLE
Drop database cleaner and use rails system specs (or as separate PR)

### ISSUE THREAD

**Issue by timdiggins** [thredded/thredded#898]
Split out from #894.

Now we are freed of pre rails 5.2 (#894), we no longer need database cleaner as we can use rails system specs.

This has multiple benefits (reliability, test flexibility, possibly speed).

**Comment by timdiggins** [comment]
We have switched to system specs: #910

However dropping database cleaner altogether is tricky: migration specs rely on it, and there is a usage of it in `user_searches_for_topics_spec.rb` which is hard to replace.

Additionally just trying to replace the remaining usages of DatabaseCleaner with use_transactional_fixtures has some occasional random failures. https://app.travis-ci.com/github/thredded/thredded/builds/246469536

So parking this for now.

**Comment by dgm** [comment]
I'm trying to drop database cleaner in a project that uses thredded, in favor of rails transactional system tests.  Tests that create thredded objects have a tendancy to crash with malloc errors, deadlocks, or double free problems.   What's different about thredded from the all the other rails gems we've used?      Any tips on doing system tests with thredded objects in our own project?

**Comment by timdiggins** [comment]
The only slightly unusual thing with Thredded is that the app uses threads to render some things in parallel (this is really a way of dealing with onebox). So maybe if your app has some problems with threading that might 

I've run Thredded-related tests in a project which uses transactional system tests successfully, including without database-cleaner

**Comment by dgm** [comment]
We don't really use threads, in production anyway, but apparently rspec is, and when I remove database cleaner and turn on transactional tests, they hang only when using thredded.   Is there a way to disable threads ?

**Comment by dgm** [comment]
Upon further reflection, this kind of has me worried - I run production in a per-process environment, not expecting threads.   Is this a possible time bomb under heavy load?

**Comment by glebm** [comment]
You can disable threaded rendering like this:

```ruby
# config/initializers/thredded.rb
Thredded::CollectionToStringsWithCacheRenderer.render_threads = 1
```

The default is 50.

Threads not working correctly indicates other issues in your application.

**Comment by dgm** [comment]
I'm sorry to hijack this thread, it's the only thing remotely close to the problem I'm seeing.   This didn't affect anything, but I got an error this time that makes me wonder about the interaction of rspec before actions, mysql, active job and thredded.   The other interesting thing here is that we don't use active job, we use delayed_job gem so I don't think we have any configuration set for active job.    It appears like active job is continuing to act on things after rspec has initiated the rollback.

`  CACHE Thredded::Messageboard Load (0.0ms)  SELECT `thredded_messageboards`.* FROM `thredded_messageboards` INNER JOIN `thredded_messageboard_commentables` ON `thredded_messageboards`.`id` = `thredded_messageboard_commentables`.`thredded_messageboard_id` WHERE `thredded_messageboard_commentables`.`commentable_id` = 1292 AND `thredded_messageboard_commentables`.`commentable_type` = 'RootCause' ORDER BY `thredded_messageboards`.`id` ASC LIMIT 1
  CACHE Thredded::Messageboard Load (0.0ms)  SELECT `thredded_messageboards`.* FROM `thredded_messageboards` INNER JOIN `thredded_messageboard_commentables` ON `thredded_messageboards`.`id` = `thredded_messageboard_commentables`.`thredded_messageboard_id` WHERE `thredded_messageboard_commentables`.`commentable_id` = 407 AND `thredded_messageboard_commentables`.`commentable_type` = 'FundingSource' ORDER BY `thredded_messageboards`.`id` ASC LIMIT 1
  CACHE Thredded::Messageboard Load (0.0ms)  SELECT `thredded_messageboards`.* FROM `thredded_messageboards` INNER JOIN `thredded_messageboard_commentables` ON `thredded_messageboards`.`id` = `thredded_messageboard_commentables`.`thredded_messageboard_id` WHERE `thredded_messageboard_commentables`.`commentable_id` = 721 AND `thredded_messageboard_commentables`.`commentable_type` = 'Monitor::InstrumentResponse' ORDER BY `thredded_messageboards`.`id` ASC LIMIT 1
  TRANSACTION (10.3ms)  ROLLBACK
[ActiveJob] [Thredded::AutoFollowAndNotifyJob] [eaaee121-44be-4c62-b2b1-eb5ae6b7df12]   Thredded::Post Load (2.2ms)  SELECT `thredded_posts`.* FROM `thredded_posts` WHERE `thredded_posts`.`id` = 120 LIMIT 1
[ActiveJob] [Thredded::AutoFollowAndNotifyJob] [eaaee121-44be-4c62-b2b1-eb5ae6b7df12] Error performing Thredded::AutoFollowAndNotifyJob (Job ID: eaaee121-44be-4c62-b2b1-eb5ae6b7df12) from Async(default) in 1050.95ms: ActiveRecord::StatementInvalid (Mysql2::Error::ConnectionError: Lost connection to MySQL server during query):
/Users/dmorton/.rvm/gems/ruby-2.7.7/gems/mysql2-0.5.4/lib/mysql2/client.rb:148:in `_query'
/Users/dmorton/.rvm/gems/ruby-2.7.7/gems/mysql2-0.5.4/lib/mysql2/client.rb:148:in `block in query'
/Users/dmorton/.rvm/gems/ruby-2.7.7/gems/mysql2-0.5.4/lib/mysql2/client.rb:147:in `handle_interrupt'
/Users/dmorton/.rvm/gems/ruby-2.7.7/gems/mysql2-0.5.4/lib/mysql2/client.rb:147:in `query'
/Users/dmorton/.rvm/gems/ruby-2.7.7/gems/activerecord-6.1.7/lib/active_record/connection_adapters/abstract_mysql_adapter.rb:206:in `block (2 levels) in execute'
/Users/dmorton/.rvm/gems/ruby-2.7.7/gems/activesupport-6.1.7/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'`

**Comment by dgm** [comment]
Yes, that led me to look at your spec helper, and I added this to my spec helper:  

```ActiveJob::Base.queue_adapter = :inline```  

And now the tests run to completion without deadlocks or other problems.

**Comment by timdiggins** [comment]
Closing this. We now mostly use transactional fixtures and only need to use DatabaseCleaner for migration specs because Mysql can't do migrations within a transaction. (see #994)
